spec/requests/uploads_controller_spec.rb

```diff
@@ -774,7 +778,577 @@ def upload_file(file, folder = "images")
+
+      it "errors if any of the parts objects have missing values" do
+        stub_list_multipart_request
+        post "/uploads/complete-multipart.json", params: {
+          unique_identifier: external_upload_stub.unique_identifier,
+          parts: [{ part_number: 1 }]
+        }
+        expect(response.status).to eq(400)
+        expect(response.body).to include("All parts must have a part number")
+      end
+
+      it "returns 404 when the upload stub does not exist" do
+        post "/uploads/complete-multipart.json", params: {
+          unique_identifier: "unknown",
+          parts: [{ part_number: 1, etag: "test1" }]
+        }
+        expect(response.status).to eq(404)
+      end
+
+      it "returns 422 when the compelte request errors" do
+        FileStore::S3Store.any_instance.stubs(:complete_multipart).raises(Aws::S3::Errors::ServiceError.new({}, "test"))
+        stub_list_multipart_request
+        post "/uploads/complete-multipart.json", params: {
+          unique_identifier: external_upload_stub.unique_identifier,
+          parts: [{ part_number: 1, etag: "test1" }]
+        }
+        expect(response.status).to eq(422)
+      end
+
+      it "returns 404 when the upload stub does not belong to the user" do
+        external_upload_stub.update!(created_by: Fabricate(:user))
+        post "/uploads/complete-multipart.json", params: {
+          unique_identifier: external_upload_stub.unique_identifier,
+          parts: [{ part_number: 1, etag: "test1" }]
+        }
+        expect(response.status).to eq(404)
+      end
+
+      it "returns 404 when the multipart upload does not exist" do
+        FileStore::S3Store.any_instance.stubs(:list_multipart_parts).raises(Aws::S3::Errors::NoSuchUpload.new("test", "test"))
+        post "/uploads/complete-multipart.json", params: {
+          unique_identifier: external_upload_stub.unique_identifier,
+          parts: [{ part_number: 1, etag: "test1" }]
+        }
+        expect(response.status).to eq(404)
+      end
+
+      it "completes the multipart upload, creates the Upload record, and returns a serialized Upload record" do
+        temp_location = "#{upload_base_url}/#{external_upload_stub.key}"
+        stub_list_multipart_request
+        stub_request(
+          :post,
+          "#{temp_location}?uploadId=#{external_upload_stub.external_upload_identifier}"
+        ).with(
+          body: "<CompleteMultipartUpload xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">\n  <Part>\n    <ETag>test1</ETag>\n    <PartNumber>1</PartNumber>\n  </Part>\n  <Part>\n    <ETag>test2</ETag>\n    <PartNumber>2</PartNumber>\n  </Part>\n</CompleteMultipartUpload>\n"
+        ).to_return(status: 200, body: <<~XML)
+<?xml version="1.0" encoding="UTF-8"?>
+<CompleteMultipartUploadResult>
+   <Location>#{temp_location}</Location>
+   <Bucket>s3-upload-bucket</Bucket>
+   <Key>#{external_upload_stub.key}</Key>
+   <ETag>testfinal</ETag>
+</CompleteMultipartUploadResult>
+        XML
```

Comment:
We're using a squishy heredoc here so I don't think we need to keep the indentation like this.